File: MicroPanda/AST/context.go
package ast

import (
	"fmt"
)

func NewContext(p *Program) *Context {
	return &Context{
		Program: p,
		objects: make(map[string]Type),
	}
}

type Context struct {
	Program  *Program
	Function *Function
	Returned bool

	parent  *Context
	objects map[string]Type
}

func (c *Context) NewContext() *Context {
	return &Context{
		Program:  c.Program,
		Function: c.Function,

		parent:  c,
		objects: make(map[string]Type),
	}
}

func (c *Context) AddObject(name string, t Type) error {
	if _, ok := c.objects[name]; ok {
		return fmt.Errorf("redeclared variable: %s", name)
	}
	c.objects[name] = t
	return nil
}

func (c *Context) FindObject(name string) Type {
	if v, ok := c.objects[name]; ok {
		return v
	}
	if c.parent != nil {
		v := c.parent.FindObject(name)
		if v != nil {
			return v
		}
	}
	if c.Function != nil && c.Function.Parent != nil {
		return c.Function.Parent.MemberType(name)
	}
	return nil
}

func (c *Context) FindSelector(selector string, member string) Type {
	parent := c.FindObject(selector)
	if parent == nil {
		_, d := c.Program.FindSelector(selector, member)
		if d == nil {
			// could be an enum
			_, e := c.Program.FindMember(selector)
			if ee, ok := e.(*Enum); ok {
				if ee.HasMember(member) {
					return TypeU8
				}
			}
			return nil
		}
		switch t := d.(type) {
		case *Variable:
			return t.Type

		case *Function:
			return t.Type

		case *Enum:
			// enum itself is not a type, its member has u8 type
			return nil
		}
	} else if t, ok := parent.(*TypeName); ok {
		d := c.Program.FindType(t)
		if s, ok := d.(*Struct); ok {
			return s.MemberType(member)
		}
	}
	return nil
}

File: MicroPanda/AST/Context.cs
namespace MicroPanda.AST;

using Declaration;

internal class Context
{
    internal Program Program { get; set; }
    internal Function? Function { get; set; }
    internal Context? Parent;
    private readonly Dictionary<string, Type.Type> objects = [];

    internal Context(Program program)
    {
        Program = program;
    }

    internal Context NewContext()
    {
        var context = new Context(Program)
        {
            Function = Function,
            Parent = this
        };
        return context;
    }

    internal bool AddObject(string name, Type.Type type)
    {
        if (objects.ContainsKey(name))
        {
            return false;
        }
        objects[name] = type;
        return true;
    }

    internal Type.Type? FindObject(string name)
    {
        if (objects.TryGetValue(name, out var type))
        {
            return type;
        }
        if (Parent != null)
        {
            type = Parent.FindObject(name);
            if (type != null)
            {
                return type;
            }
        }
        if (Function != null && Function.Parent != null)
        {
            return Function.Parent.MemberType(name);
        }
        return null;
    }

    public Type.Type? FindSelector(string selector, string member)
    {
        var parent = FindObject(selector);
        if (parent == null)
        {
            var declaration = Program.FindSelector(selector, member);
            if (declaration == null)
            {
                // could be an enum
                var e = Program.FindMember(selector);
                if (e is Enum ee && ee.HasMember(member))
                {
                    return Type.TypeHelper.TypeU8;
                }
                return null;
            }
            switch (declaration)
            {
                case Variable variable:
                    return variable.Type;

                case Function function:
                    return function.Type;

                case Enum _:
                    // enum itself is not a type, its member has u8 type
                    return null;
            }
        }
        else if (parent is Type.TypeName typeName)
        {
            var declaration = Program.FindByTypeName(typeName);
            if (declaration is Struct s)
            {
                return s.MemberType(member);
            }
        }
        return null;
    }
}
